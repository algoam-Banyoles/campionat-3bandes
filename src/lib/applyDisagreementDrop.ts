import type { SupabaseClient } from '@supabase/supabase-js';

export async function applyDisagreementDrop(
  supabase: SupabaseClient,
  eventId: string,
  playerA: string,
  playerB: string
): Promise<void> {
  const { error } = await supabase.rpc('apply_disagreement_drop', {
    event_id: eventId,
    player_a: playerA,
    player_b: playerB
  });
  if (error) throw new Error(error.message);
}
