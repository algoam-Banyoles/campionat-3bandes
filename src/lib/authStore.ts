import { writable } from 'svelte/store';
import type { Session, User } from '@supabase/supabase-js';
import { supabase } from '$lib/supabaseClient';

export const session = writable<Session | null>(null);
export const user = writable<User | null>(null);
export const loadingAuth = writable(true);

async function init() {
  const { data } = await supabase.auth.getSession();
  session.set(data.session ?? null);
  user.set(data.session?.user ?? null);
  loadingAuth.set(false);

  supabase.auth.onAuthStateChange((_event, sess) => {
    session.set(sess ?? null);
    user.set(sess?.user ?? null);
  });
}
init();
