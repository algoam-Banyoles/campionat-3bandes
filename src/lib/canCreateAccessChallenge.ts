import type { SupabaseClient } from '@supabase/supabase-js';

export type CanCreateAccessResult = {
  ok: boolean;
  reason: string | null;
};

export async function canCreateAccessChallenge(
  supabase: SupabaseClient,
  eventId: string,
  reptadorId: string,
  reptatId: string
): Promise<CanCreateAccessResult> {
  const { data, error } = await supabase.rpc('can_create_access_challenge', {
    p_event: eventId,
    p_reptador: reptadorId,
    p_reptat: reptatId
  });
  if (error) return { ok: false, reason: error.message };
  const result = data as CanCreateAccessResult[] | null;
  if (!result || result.length === 0) return { ok: false, reason: 'No result' };
  return result[0];
}
