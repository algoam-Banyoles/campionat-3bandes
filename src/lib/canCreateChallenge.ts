import type { SupabaseClient } from '@supabase/supabase-js';

export type CanCreateChallengeResult = {
  ok: boolean;
  reason: string | null;
  warning: string | null;
};

export async function canCreateChallenge(
  supabase: SupabaseClient,
  eventId: string,
  reptadorId: string,
  reptatId: string
): Promise<CanCreateChallengeResult> {
  const { data, error } = await supabase.rpc('can_create_challenge', {
    p_event: eventId,
    p_reptador: reptadorId,
    p_reptat: reptatId
  });
  if (error) return { ok: false, reason: error.message, warning: null };
  const result = data as CanCreateChallengeResult[] | null;
  if (!result || result.length === 0) return { ok: false, reason: 'No result', warning: null };
  return result[0];
}
