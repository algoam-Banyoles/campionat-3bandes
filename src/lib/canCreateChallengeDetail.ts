import type { SupabaseClient } from '@supabase/supabase-js';

export type CanCreateChallengeDetailResult = {
  ok: boolean;
  reason?: string | null;
};

export async function canCreateChallengeDetail(
  supabase: SupabaseClient,
  eventId: string,
  reptadorId: string,
  reptatId: string
): Promise<CanCreateChallengeDetailResult> {
  const { data, error } = await supabase.rpc('can_create_challenge_detail', {
    event_id: eventId,
    reptador_id: reptadorId,
    reptat_id: reptatId
  });
  if (error) {
    return { ok: false, reason: error.message };
  }
  // Supabase RPC may return single object or array
  const result = Array.isArray(data) ? data[0] : data;
  if (!result) {
    return { ok: false, reason: 'No result' };
  }
  return result as CanCreateChallengeDetailResult;
}

