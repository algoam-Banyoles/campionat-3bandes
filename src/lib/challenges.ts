import type { SupabaseClient } from '@supabase/supabase-js';
import { authFetch } from '$lib/utils/http';

export type ChallengeParticipants = {
  reptador_id: string | null;
  reptat_id: string | null;
};

export function isParticipant(
  mePlayerId: string | null,
  repte: ChallengeParticipants | null | undefined
): boolean {
  if (!mePlayerId || !repte) return false;
  return mePlayerId === repte.reptador_id || mePlayerId === repte.reptat_id;
}

export async function acceptChallenge(supabase: SupabaseClient, id: string): Promise<void> {
  const { error } = await supabase
    .from('challenges')
    .update({ estat: 'acceptat', data_acceptacio: new Date().toISOString() })
    .eq('id', id);
  if (error) throw new Error(error.message);
}

export async function refuseChallenge(supabase: SupabaseClient, id: string): Promise<void> {
  const { error } = await supabase
    .from('challenges')
    .update({ estat: 'refusat' })
    .eq('id', id);
  if (error) throw new Error(error.message);
}

export async function scheduleChallenge(
  _supabase: SupabaseClient,
  id: string,
  isoDate: string
): Promise<void> {
  const res = await authFetch('/campionat-continu/reptes/programar', {
    method: 'POST',
    body: JSON.stringify({ id, data_iso: isoDate })
  });
  const body = await res.json().catch(() => ({}));
  if (!res.ok || body?.ok === false) {
    throw new Error(body?.error || 'Error programant repte');
  }
}


export async function resolveAccessChallenge(
  supabase: SupabaseClient,
  challengeId: string,
  winner: 'reptador' | 'reptat'
): Promise<{ ok: boolean; error?: string }> {
  const { data, error } = await supabase.rpc('resolve_access_challenge', {
    challenge_id: challengeId,
    winner
  });
  if (error) throw new Error(error.message);
  const res = (data as any) ?? { ok: false };
  if (!res.ok) throw new Error(res.error || 'Error resolent repte d\'acc√©s');
  return res;
}

