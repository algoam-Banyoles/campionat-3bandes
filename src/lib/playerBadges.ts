export type PlayerBadgeSource = {
  player_id: string;
  isMe?: boolean;
  hasActiveChallenge?: boolean;
  cooldownToChallenge?: boolean;
  reptable?: boolean;
  canChallenge?: boolean;
  reason?: string | null;
  protected?: boolean;
  outside?: boolean;
};

export type PlayerBadgeState =
  | { type: 'self' }
  | { type: 'active-challenge' }
  | { type: 'cooldown' }
  | { type: 'can-challenge' }
  | { type: 'reptable'; canChallenge: boolean; reason: string | null }
  | { type: 'protected' }
  | { type: 'outside' };

export function getPlayerBadges(row: PlayerBadgeSource): PlayerBadgeState[] {
  const badges: PlayerBadgeState[] = [];

  if (row.isMe) {
    badges.push({ type: 'self' });
  }

  if (row.hasActiveChallenge) {
    badges.push({ type: 'active-challenge' });
  } else if (row.cooldownToChallenge) {
    badges.push({ type: 'cooldown' });
  } else if (row.isMe) {
    badges.push({ type: 'can-challenge' });
  } else if (row.reptable) {
    badges.push({
      type: 'reptable',
      canChallenge: !!row.canChallenge,
      reason: row.reason ?? null
    });
  }

  if (row.protected) {
    badges.push({ type: 'protected' });
  }

  if (row.outside) {
    badges.push({ type: 'outside' });
  }

  return badges;
}

export type PlayerBadgeDescriptor = {
  state: PlayerBadgeState['type'];
  element: 'span' | 'button';
  className: string;
  text: string;
  title: string;
  ariaLabel: string;
  disabled?: boolean;
};

const CLASS_BY_TYPE: Record<PlayerBadgeState['type'], string> = {
  self:
    'ml-1 inline-block rounded-full bg-yellow-400 px-2.5 py-1 text-xs font-medium text-gray-900 align-middle',
  'active-challenge':
    'ml-1 inline-block rounded-full bg-red-600 px-2.5 py-1 text-xs font-medium text-white align-middle',
  cooldown:
    'ml-1 inline-block rounded-full bg-yellow-300 px-2.5 py-1 text-xs font-medium text-gray-900 align-middle',
  'can-challenge':
    'ml-1 inline-block rounded-full bg-green-600 px-2.5 py-1 text-xs font-medium text-white align-middle',
  reptable:
    'ml-1 rounded-full bg-blue-600 px-2.5 py-1 text-xs font-medium text-white align-middle disabled:opacity-50',
  protected:
    'ml-1 inline-block rounded-full bg-gray-400 px-2.5 py-1 text-xs font-medium text-white align-middle',
  outside:
    'ml-1 inline-block rounded-full bg-gray-200 px-2.5 py-1 text-xs font-medium text-gray-800 align-middle'
};

const TEXT_BY_TYPE: Record<PlayerBadgeState['type'], string> = {
  self: 'Tu',
  'active-challenge': 'Té repte actiu',
  cooldown: 'No pot reptar',
  'can-challenge': 'Pot reptar',
  reptable: 'Reptable',
  protected: 'Protegit',
  outside: 'Fora de rànquing actiu'
};

const TITLE_BY_TYPE = (
  badge: PlayerBadgeState
): string => {
  switch (badge.type) {
    case 'self':
      return 'Tu';
    case 'active-challenge':
      return 'Aquest jugador té un repte actiu (no pot iniciar-ne cap altre).';
    case 'cooldown':
      return 'En cooldown fins passats 7 dies del darrer repte disputat.';
    case 'can-challenge':
      return 'Pots reptar fins a 2 posicions per sobre teu.';
    case 'reptable':
      return badge.canChallenge
        ? 'Aquest jugador és reptable per tu ara mateix.'
        : badge.reason ?? 'No pots reptar';
    case 'protected':
      return 'Protegit (cooldown de ser reptat)';
    case 'outside':
      return 'Fora de rànquing actiu';
  }
  return '';
};

export function buildBadgeDescriptors(
  row: PlayerBadgeSource,
  provider: (source: PlayerBadgeSource) => PlayerBadgeState[] = getPlayerBadges
): PlayerBadgeDescriptor[] {
  return provider(row).map((badge) => {
    const title = TITLE_BY_TYPE(badge);
    return {
      state: badge.type,
      element: badge.type === 'reptable' ? 'button' : 'span',
      className: CLASS_BY_TYPE[badge.type],
      text: TEXT_BY_TYPE[badge.type],
      title,
      ariaLabel: title,
      disabled: badge.type === 'reptable' ? !badge.canChallenge : undefined
    } satisfies PlayerBadgeDescriptor;
  });
}
