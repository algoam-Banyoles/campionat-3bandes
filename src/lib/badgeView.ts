import { getPlayerBadges, type VPlayerBadges } from '$lib/playerBadges';

export type BadgeView = {
  label: string;
  text: string;
  className: string;
};

const ACTIVE_BADGE: BadgeView = {
  label: 'Repte actiu',
  text: 'Repte actiu',
  className: 'rounded bg-blue-100 px-1.5 py-0.5 text-xs font-medium text-blue-700'
};

const CHALLENGEABLE_BADGE: BadgeView = {
  label: 'Es pot reptar',
  text: 'Es pot reptar',
  className: 'rounded bg-green-100 px-1.5 py-0.5 text-xs font-medium text-green-700'
};

export function getBadgeView(badge: VPlayerBadges | undefined | null): BadgeView | null {
  if (!badge) return null;
  if (badge.has_active_challenge) {
    return ACTIVE_BADGE;
  }
  if (badge.in_cooldown) {
    return null;
  }
  if (badge.can_be_challenged) {
    return CHALLENGEABLE_BADGE;
  }
  return null;
}

export async function fetchBadgeMap(
  getBadges: () => Promise<VPlayerBadges[]> = getPlayerBadges
): Promise<Map<string, VPlayerBadges>> {
  const list = await getBadges();
  return new Map(list.map((badge) => [badge.player_id, badge]));
}
