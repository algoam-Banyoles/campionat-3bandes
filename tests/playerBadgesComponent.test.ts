import { describe, it, expect, vi } from 'vitest';
import type { PlayerBadgeState } from '../src/lib/playerBadges';
import { buildBadgeDescriptors } from '../src/lib/playerBadges';

const baseRow = { player_id: 'player-1' };

const runWith = (states: PlayerBadgeState[]) => {
  const provider = vi.fn().mockReturnValue(states);
  const badges = buildBadgeDescriptors(baseRow as any, provider);
  expect(provider).toHaveBeenCalledWith(baseRow);
  return badges;
};

describe('PlayerBadges descriptors', () => {
  it('creates active challenge badge descriptor', () => {
    const [badge] = runWith([{ type: 'active-challenge' }]);
    expect(badge.element).toBe('span');
    expect(badge.text).toBe('Té repte actiu');
    expect(badge.ariaLabel).toBe('Aquest jugador té un repte actiu (no pot iniciar-ne cap altre).');
    expect(badge.className).toContain('bg-red-600');
  });

  it('creates cooldown badge descriptor', () => {
    const [badge] = runWith([{ type: 'cooldown' }]);
    expect(badge.text).toBe('No pot reptar');
    expect(badge.ariaLabel).toBe('En cooldown fins passats 7 dies del darrer repte disputat.');
    expect(badge.className).toContain('bg-yellow-300');
  });

  it('creates can-challenge descriptor for self', () => {
    const [badge] = runWith([{ type: 'can-challenge' }]);
    expect(badge.text).toBe('Pot reptar');
    expect(badge.ariaLabel).toBe('Pots reptar fins a 2 posicions per sobre teu.');
    expect(badge.className).toContain('bg-green-600');
  });

  it('creates enabled reptable descriptor', () => {
    const states: PlayerBadgeState[] = [{ type: 'reptable', canChallenge: true, reason: 'Disponible' }];
    const [badge] = runWith(states);
    expect(badge.element).toBe('button');
    expect(badge.text).toBe('Reptable');
    expect(badge.ariaLabel).toBe('Aquest jugador és reptable per tu ara mateix.');
    expect(badge.disabled).toBe(false);
  });

  it('creates disabled reptable descriptor', () => {
    const states: PlayerBadgeState[] = [{ type: 'reptable', canChallenge: false, reason: 'No pots reptar' }];
    const [badge] = runWith(states);
    expect(badge.disabled).toBe(true);
    expect(badge.ariaLabel).toBe('No pots reptar');
  });

  it('creates protected descriptor', () => {
    const [badge] = runWith([{ type: 'protected' }]);
    expect(badge.text).toBe('Protegit');
    expect(badge.ariaLabel).toBe('Protegit (cooldown de ser reptat)');
    expect(badge.className).toContain('bg-gray-400');
  });

  it('creates outside descriptor', () => {
    const [badge] = runWith([{ type: 'outside' }]);
    expect(badge.text).toBe('Fora de rànquing actiu');
    expect(badge.ariaLabel).toBe('Fora de rànquing actiu');
    expect(badge.className).toContain('bg-gray-200');
  });

  it('creates self descriptor', () => {
    const [badge] = runWith([{ type: 'self' }]);
    expect(badge.text).toBe('Tu');
    expect(badge.ariaLabel).toBe('Tu');
    expect(badge.className).toContain('bg-yellow-400');
  });

  it('matches snapshot for multiple badges', () => {
    const states: PlayerBadgeState[] = [
      { type: 'self' },
      { type: 'active-challenge' },
      { type: 'protected' }
    ];
    const provider = vi.fn().mockReturnValue(states);
    const badges = buildBadgeDescriptors(baseRow as any, provider).map(({ state, element, text, ariaLabel, className, disabled }) => ({
      state,
      element,
      text,
      ariaLabel,
      className,
      disabled: disabled ?? false
    }));
    expect(badges).toMatchSnapshot();
  });
});
