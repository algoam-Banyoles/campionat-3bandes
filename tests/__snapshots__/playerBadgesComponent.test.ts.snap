// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlayerBadges descriptors > matches snapshot for multiple badges 1`] = `
[
  {
    "ariaLabel": "Tu",
    "className": "ml-1 inline-block rounded-full bg-yellow-400 px-2.5 py-1 text-xs font-medium text-gray-900 align-middle",
    "disabled": false,
    "element": "span",
    "state": "self",
    "text": "Tu",
  },
  {
    "ariaLabel": "Aquest jugador té un repte actiu (no pot iniciar-ne cap altre).",
    "className": "ml-1 inline-block rounded-full bg-red-600 px-2.5 py-1 text-xs font-medium text-white align-middle",
    "disabled": false,
    "element": "span",
    "state": "active-challenge",
    "text": "Té repte actiu",
  },
  {
    "ariaLabel": "Protegit (cooldown de ser reptat)",
    "className": "ml-1 inline-block rounded-full bg-gray-400 px-2.5 py-1 text-xs font-medium text-white align-middle",
    "disabled": false,
    "element": "span",
    "state": "protected",
    "text": "Protegit",
  },
]
`;
